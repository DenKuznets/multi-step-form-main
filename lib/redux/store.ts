import { configureStore } from '@reduxjs/toolkit';
import appReducer from './slices/appSlice';
// ...
/* eslint-disable no-underscore-dangle */

export const store = configureStore({
    reducer: {
        app: appReducer
    }
});
/* eslint-enable */
// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
