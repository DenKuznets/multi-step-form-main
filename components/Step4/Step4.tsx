'use client';
import { useRouter } from 'next/navigation';
import Link from 'next/link';
import { useAppDispatch, useAppSelector } from '@/lib/redux/hooks';
import {
    selectAddons,
    selectPaymentMethod,
    selectPlan,
    setValid
} from '@/lib/redux/slices/appSlice';
import { Addons, PAYMENT, Plans } from '@/utils/steps';
import { useForm } from 'react-hook-form';

const Step4 = () => {
    const dispatch = useAppDispatch();
    const currentPaymentMethod = useAppSelector(selectPaymentMethod);
    const currentPlanName = useAppSelector(selectPlan);
    const currentPlan = Plans.find((plan) => plan.name === currentPlanName);
    const currentAddonsNames = useAppSelector(selectAddons);
    const currentAddons = Addons.filter((addon) =>
        currentAddonsNames.includes(addon.name)
    );

    const { handleSubmit } = useForm();

    const router = useRouter();
    const monthly = currentPaymentMethod === PAYMENT.MONTHLY;
    const period = monthly ? 'mo' : 'yr';
    let totalPrice = currentPlan
        ? monthly
            ? currentPlan.priceMonth
            : currentPlan.priceYear
        : 0;

    const onSubmit = () => {
        dispatch(setValid(true));
    };

    const addonsList = currentAddons.map((addon) => {
        totalPrice += monthly ? addon.priceMonth : addon.priceYear;
        return (
            <div key={addon.name} className="flex justify-between">
                <span>{addon.header}</span>
                <span className="text-marineBlue">
                    +${monthly ? `${addon.priceMonth}` : `${addon.priceYear}`}/
                    {period}
                </span>
            </div>
        );
    });

    return (
        <form
            noValidate
            aria-label="select summary form"
            className="text-sm text-coolGray [&_*]:first-letter:capitalize"
            onSubmit={handleSubmit(onSubmit)}
        >
            <div className="mb-6 bg-alabaster p-4 lg:p-6">
                {currentPlan ? (
                    <div className="flex items-center justify-between">
                        <span>
                            <span className="font-bold capitalize text-marineBlue lg:text-base">
                                {currentPlan.name} (
                                {monthly ? 'monthly' : 'yearly'})
                            </span>
                            <br />
                            <Link
                                className="capitalize underline transition-all hover:text-purplishBlue"
                                href={'/select-plan'}
                            >
                                change
                            </Link>
                        </span>
                        <span className="font-bold text-marineBlue lg:text-base">
                            $
                            {monthly
                                ? `${currentPlan.priceMonth}`
                                : `${currentPlan.priceYear}`}
                            /{period}
                        </span>
                    </div>
                ) : (
                    <Link href={'/select-plan'}>Select plan</Link>
                )}

                <hr className="my-3 lg:mb-4 lg:mt-6" />
                <div className="flex flex-col gap-3 lg:gap-4">{addonsList}</div>
            </div>
            <div className="flex justify-between px-4 lg:px-6">
                <span>total (per {monthly ? 'month' : 'year'})</span>
                <span className="text-base font-bold text-blue-700 lg:text-xl">
                    ${totalPrice}/{period}
                </span>
            </div>
            <div className="fixed bottom-0  left-0 flex w-full justify-between bg-white p-4 shadow sm:absolute sm:mt-8 sm:p-0 sm:shadow-none md:mb-4">
                <button
                    onClick={(e) => {
                        e.preventDefault();
                        router.push('/add-ons');
                    }}
                    className="btn btn-back text-coolGray"
                >
                    Go Back
                </button>
                <button className="btn h-10 w-24 rounded-[4px] bg-blue-700 text-white transition-all hover:bg-blue-500 lg:h-12 lg:w-32 lg:rounded-lg">
                    Confirm
                </button>
            </div>
        </form>
    );
};

export default Step4;
