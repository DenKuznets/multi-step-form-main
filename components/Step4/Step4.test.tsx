import '@testing-library/jest-dom';
import { renderWithProviders, screen } from '@/lib/redux/utils-for-tests';
import Step4 from './Step4';

jest.mock('next/navigation', () => ({
    useRouter() {
        return {
            prefetch: () => null
        };
    }
}));

test('renders correctly', () => {
    renderWithProviders(<Step4 />);

    const form = screen.getByRole('form', { name: /select summary form/i });

    const backbutton = screen.getByRole('button', {
        name: /go back/i
    });
    const confirmBtn = screen.getByRole('button', {
        name: /confirm/i
    });

    expect(form).toBeInTheDocument();
    expect(backbutton).toBeInTheDocument();
    expect(confirmBtn).toBeInTheDocument();
});

test('snapshot Step4 unchanged', () => {
    const { container } = renderWithProviders(<Step4 />);
    expect(container).toMatchSnapshot();
});
