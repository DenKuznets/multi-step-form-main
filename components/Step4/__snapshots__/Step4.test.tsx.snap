// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot Step4 unchanged 1`] = `
<div>
  <form
    aria-label="select summary form"
    class="text-sm text-coolGray [&_*]:first-letter:capitalize"
    novalidate=""
  >
    <div
      class="mb-6 bg-alabaster p-4 lg:p-6"
    >
      <div
        class="flex items-center justify-between"
      >
        <span>
          <span
            class="font-bold capitalize text-marineBlue lg:text-base"
          >
            arcade
             (
            monthly
            )
          </span>
          <br />
          <a
            class="capitalize underline transition-all hover:text-purplishBlue"
            href="/select-plan"
          >
            change
          </a>
        </span>
        <span
          class="font-bold text-marineBlue lg:text-base"
        >
          $
          9
          /
          mo
        </span>
      </div>
      <hr
        class="my-3 lg:mb-4 lg:mt-6"
      />
      <div
        class="flex flex-col gap-3 lg:gap-4"
      />
    </div>
    <div
      class="flex justify-between px-4 lg:px-6"
    >
      <span>
        total (per 
        month
        )
      </span>
      <span
        class="text-base font-bold text-blue-700 lg:text-xl"
      >
        $
        9
        /
        mo
      </span>
    </div>
    <div
      class="form-btns-container"
    >
      <button
        class="btn btn-back text-coolGray"
      >
        Go Back
      </button>
      <button
        class="btn h-10 w-24 rounded-[4px] bg-blue-700 text-white transition-all hover:bg-blue-500 lg:h-12 lg:w-32 lg:rounded-lg"
      >
        Confirm
      </button>
    </div>
  </form>
</div>
`;
